lightningd/invoice.c

```diff
@@ -265,6 +266,8 @@ void invoice_try_pay(struct lightningd *ld,
 	payload->hin = hin;
 	tal_add_destructor2(hin, invoice_payload_remove_hin, payload);
 
+	notify_invoice_payment(ld, payload->msat, payload->preimage, payload->label);
+
 	log_debug(ld->log, "Calling hook for invoice '%s'", details->label->s);
 	plugin_hook_call_invoice_payment(ld, payload, payload);
```

Comment:
Should we perform the notify * before* the plugin-hook?

The plugin-hook may (in theory) tell the `lightningd` to reject the payment, so we would end up notifying an invoice payment that does not push through.  Would not it be better to notify *after* the plugin-hook returns with an acceptance?

Reply:
Actually that was exactly why I put the hook call before in the first place.

Since, as pointed out by cdecker above, we don't have ownership on `payload` when passed to notify after the hook. To avoid a copy I assumed a plugin which has subsribed to the `invoice_payment` hook won't rely on the (or even subscribe to)  `invoice_payment` notification to know if a payment arrived.

Reply:
Maybe we can put the notification call into the hook callback? That should get back a copy of the payload as well.

Reply:
> Maybe we can put the notification call into the hook callback? That should get back a copy of the payload as well.

Wouldn't that condition the notification to be sent only if the hook is registered too ?

Reply:
No, the hook callback is always called.  If there is no plugin registered, it gets a `buffer` of `NULL`, which it should always check for.